subsys/net/lib/sockets/sockets_select.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sys/time.h>
+
+#include <net/socket.h>
+
+static int *fdset_find_fd(int fd, zsock_fd_set *set)
+{
+	int i, left = set->active;
+
+	for (i = 0; left > 0 && i < ARRAY_SIZE(set->fds); i++) {
+		if (set->fds[i] == fd) {
+			return &set->fds[i];
+		}
+
+		if (set->fds[i] != -1) {
+			left--;
+		}
+	}
+
+	return NULL;
+}
+
+void ZSOCK_FD_ZERO(zsock_fd_set *set)
+{
+	int i;
+
+	set->active = 0;
+
+	for (i = 0; i < ARRAY_SIZE(set->fds); i++) {
+		set->fds[i] = -1;
+	}
+}
+
+int ZSOCK_FD_ISSET(int fd, zsock_fd_set *set)
+{
+	int *p = fdset_find_fd(fd, set);
+
+	return p != NULL;
+}
+
+void ZSOCK_FD_CLR(int fd, zsock_fd_set *set)
+{
+	int *p = fdset_find_fd(fd, set);
+
+	if (p != NULL) {
+		*p = -1;
+		set->active--;
+	}
+}
+
+void ZSOCK_FD_SET(int fd, zsock_fd_set *set)
+{
+	int i;
+
+	if (set->active == ARRAY_SIZE(set->fds)) {
+		/* No room, no way to indicate that. */
+		return;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(set->fds); i++) {
+		if (set->fds[i] == -1) {
+			set->fds[i] = fd;
+			set->active++;
+			return;
+		}
+	}
+}
+
+static int fd_set_to_pollfd(int poll_flags, zsock_fd_set *set,
+			    struct pollfd *pfds, int *pfds_idx)
+{
+	int i, left;
+	int poll_idx = *pfds_idx;
+
+	if (set == NULL) {
+		return 0;
+	}
+
+	if (set->active > (CONFIG_NET_SOCKETS_POLL_MAX - poll_idx)) {
+		errno = ENOMEM;
+		return -1;
+	}
+
+	for (i = 0, left = set->active; i > 0 && left > 0; i++) {
+		if (set->fds[i] != -1) {
+			pfds[poll_idx].fd = set->fds[i];
```

Comment:
Shouldn't we check if an entry for specific file descriptor already exists in the `pfds`? Currently the function will create a separate entry if the specific fd is present in more than one set. While it still should work, I'm not sure if it is what we want.

Reply:
> While it still should work, I'm not sure if it is what we want.

Well, we want to implement it quickly and get it off the list, so yeah, the bet is on "it still should work" ;-).

That said, as I mentioned, I kinda morally ready to rework this impl using bitsets, which would take care of this issue too.